### PR TITLE
Scopes with foreign key

### PR DESCRIPTION
**Pull request by u9r52sld**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/master/CONTRIBUTING.md)?

### Description of change

Confirming options after foreign key is added.

See issue: #6245

### CODE REVIEW

**Comment by sushantdhiman** [comment]
Thanks for this fix. I want to do some tests on my local setup after that I will merge this PR

**Comment by sushantdhiman** [comment]
Tested this locally, this wont work with `defaultScope` but I have no good suggestion to handle that case right now. Apart from that it works well with other scopes defined by `options.scope`

@u9r52sld I can merge this in current state or would you like to fix `defaultScope` case as well?

**Comment by u9r52sld** [comment]
@sushantdhiman Thank you for testing. Of course I'll fix `defaultScope` case. Please hold on.

**Comment by u9r52sld** [comment]
Someone please tell me why the `codecov/project` failed. What should we do?

**Comment by sushantdhiman** [comment]
> Someone please tell me why the codecov/project failed. What should we do?

You can ignore it in this case.

LGTM, I'll again test a few things locally and merge

**Comment by sushantdhiman** [comment]
Finally got the chance to test this (basically was not active with Sequelize for past week). Looks like attributes mixing is broken after this, consider this case

```js
const Model = sequelize.define('model', {
  name: Sequelize.STRING,
  value: Sequelize.INTEGER,
  password: Sequelize.STRING
}, {
    defaultScope: {
      attributes: {
        exclude: ['password']
      }
    },
    scopes: {
      aScope: {
        attributes: {
          exclude: ['value']
        }
      }
    }
  });

console.log(
  Model._scope
);

console.log(
  Model.scope('aScope', 'defaultScope')._scope
);
```

```
{ attributes: { exclude: [ 'password' ] } } // correct
{ attributes: { exclude: [ 'value' ] } } // incorrect, exclude should include both `password`, `value`
```

**Comment by sushantdhiman** [comment]
I think we should conform scope just before merging them that will solve all concerns I have raised so far. Just conform scopes before merging with other scopes and keep merge/assign logic same as before (as much as possible), otherwise it can become an unintended breaking change.

**Comment by u9r52sld** [comment]
> I think we should conform scope just before merging them that will solve all concerns I have raised so far

Could you please tell me the reason?

I think we should not conform scopes like `exclude` or `include` before queries are executed because it can be added new `foreign_key` fields by associations after that.

**Comment by sushantdhiman** [comment]
May be we can merge attributes in that style, but not for `include`.

Conforming will convert both `include: [{ model: ABC }]` and `include: [ABC]` to same format so we don't have duplicate associations.

Conform standardise includes and other properties of scope so merge is predictable

**Comment by sushantdhiman** [comment]
@u9r52sld Thanks, now it looks almost ready.

Some concerns related to merge strategy changes, we have discussed it in https://github.com/sequelize/sequelize/issues/9087 for a few months. I have asked participants from that discussion to share their ideas here. I don't use scopes in my projects so not sure what merge strategy works well. I hope we can figure out something that works for everyone.

P.S. I'll be out for a few days, after that we can complete our discussion and merge this to v5

**Comment by papb** [comment]
Hello, I'm the one who opened #9087, I believe sushantdhiman wanted me to take a look here because this is also related to scopes. I will try to take a look as soon as possible (in less than 48 hours probably). In the meantime, if you could read my concerns on #9087 as well that would be great :) I'll be happy to discuss further and answer questions.

**Comment by papb** [comment]
Hello again, I took a closer look, this PR seems tangentially related to my issue #9087 because it is also related to scopes, but if I'm not mistaken that's about it. @u9r52sld, what do you think? If you could take a look as well that would be great.

**Comment by sushantdhiman** [comment]
@u9r52sld @papb wrt to changes in this PR we should only target how scopes are merged among themselves. That is under scope of this PR.

For overall scope overhaul next we need to decide how scopes are merged with finder options, but that should be a different PR.

This PR is changing both scope's internal representation (like attributes are no longer converted from `include/exclude` to plain format etc) and strategy how to merge with other scopes. I want your suggestion (possibly others may be), if this current merge strategy (for merging scopes among themselves) is ok, or need any more changes (which btw looks good to me)

**Comment by papb** [comment]
Oh, the way that scopes are merged among themselves does concern my issue as well.

Currently, two scopes with different includes do not merge correctly. I show this in my issue #9087.

Once I have more time, I can prepare an specific test that fails. But I believe this should be straightforward from my issue.

If anyone has questions let me know. In the weekend I'll have time to write the failing test.

**Comment by papb** [comment]
Hello, I didn't have time to write the test as I was thinking, but please take a look at [this other test that I wrote](https://github.com/sequelize/sequelize/pull/9252#issuecomment-405801477). It is about a merge of a scope with the findAll options, but the same idea for merging two scopes fails the exact same way.

Let me know if you have any other questions or want clarifications.

**Comment by u9r52sld** [comment]
@sushantdhiman @papb I'm sorry for the late reply. Thank you for your concern!
I'll check issue #9087 and fix it.

**Comment by papb** [comment]
@u9r52sld Thank you very much! Let me know if you need any further clarifications.

File: test/unit/model/scope.test.js [hunk 1]
```diff
@@ -310,15 +305,15 @@ describe(Support.getTestDialectTeaser('Model'), () => {
...
+        include: [{ model: Project }]
       }, { override: true });
 
       expect(Company._scope).to.deep.equal({
         include: [{ model: Project }]
       });
     });
 
-    it('works with exclude and include attributes', () => {
+    it('should not work with exclude and include attributes', () => {
```
Comment (sushantdhiman) [inline test/unit/model/scope.test.js]: `should keep exclude and include attributes`

File: lib/model.js [hunk 2]
```diff
@@ -3052,35 +3052,19 @@ class Model {
...
+    Utils.defaults(options, scope);
+    Utils.defaults(options.where, scope.where);
 
+    // Includes need special treatment
     if (scope.include) {
-      options.include = options.include || [];
-
-      // Reverse so we consider the latest include first.
-      // This is used if several scopes specify the same include - the last scope should take precedence
-      for (const scopeInclude of scope.include.reverse()) {
```
Comment (sushantdhiman) [inline lib/model.js]: What is the new merge strategy now (wrt https://github.com/sequelize/sequelize/blob/master/docs/scopes.md#merging), as far as I can tell it is still the same?

Comment (u9r52sld) [reply lib/model.js]: Yeah, it's almost the same. I just updated for issues #9087,#6245

For example, current merge strategy is as follows.

```
Company.addScope('scopeA', {
  where: { name: 'A', something: true },
  attributes: {
    include: ['name']
  },
  include: [
    {
      model: Project,
      where: { name: 'A', something: true },
    },
    Project
  ],
  limit: 5,
  order: [
    ['name', 'DESC'],
  ]
})

Company.addScope('scopeB', {
  where: { name: 'B' },
  attributes: {
    include: ['something']
  },
  include: [
    {
      model: Project,
      as: 'Pj'
    },
    {
      model: Project,
      where: { name: 'B' }
    },
  ],
  limit: 10,
  order: [
    ['something', 'ASC'],
  ]
})

expect(Company.scope(['scopeA', 'scopeB'])._scope).to.deep.equal({
  where: { name: 'B', something: true }, // 'where' is shallowly merged
  attributes: { // just override
    include: ['something']
  },
  include: [ // 'include' is shallowly merged with the same model and alias(as)
    {
      model: Project,
      where: { name: 'B', something: true },
    },
    {
      model: Project,
      as: 'Pj'
    },
  ],
  limit: 10, // just override
  order: [ // 'Array' is united
    ['name', 'DESC'],
    ['something', 'ASC']
  ]
})
```

Comment (sushantdhiman) [reply lib/model.js]: Hmm I thought we discussed to have `attributes.include / exclude` merge https://github.com/sequelize/sequelize/pull/9735#issuecomment-413088477

Comment (u9r52sld) [reply lib/model.js]: Oh, i'm sorry. I fixed it. e31be36

```
Company.addScope('scopeA', {
  attributes: {
    include: ['name']
  }
})

Company.addScope('scopeB', {
  attributes: {
    include: ['something'],
    exclude: ['password']
  }
})

expect(Company.scope(['scopeA', 'scopeB'])._scope).to.deep.equal({
  attributes: { // merged
    include: ['name', 'something'],
    exclude: ['password']
  }
})
```

**Comment by papb** [comment]
This is looking very good :+1:

Thanks @u9r52sld! I tested your changes against my examples in #9087 and I'm very happy with the results. I also tested more complicated and nested combinations and they also worked as I'd like. Thank you very much! :grin:

**Comment by sushantdhiman** [comment]
Thanks @u9r52sld for delvering after this 3 months long review process :) & thanks @papb for your additional comments and review

**Comment by harpcio** [comment]
Are there any chances for this bugfix to be in v4?

**Comment by papb** [comment]
Extra documentation for this PR was added in #10312

**Pull request merged**